patroni/quorum.py

```diff
@@ -0,0 +1,407 @@
+    :ivar leader_wanted: the desired leader (could be different from the *leader* right after a failover).
+    """
+
+    def __init__(self, leader: str, quorum: int, voters: Collection[str],
+                 numsync: int, sync: Collection[str], numsync_confirmed: int,
+                 active: Collection[str], sync_wanted: int, leader_wanted: str) -> None:
+        """Instantiate :class:``QuorumStateResolver`` based on input parameters.
+
+        :param leader: name of the leader, according to the ``/sync`` key.
+        :param quorum: ``quorum`` value from the ``/sync`` key, the minimal number of nodes we need see
+                        when doing the leader race.
+        :param voters: ``sync_standby`` value from the ``/sync`` key, set of node names we will be
+                       running the leader race against.
+        :param numsync: the number of synchronous nodes from the ``synchronous_standby_names``.
+        :param sync: Set of node names listed in the ``synchronous_standby_names``.
+        :param numsync_confirmed: the number of nodes that are confirmed to reach "safe" LSN after
+                                  they were added to the ``synchronous_standby_names``.
+        :param active: set of node names that are replicating from the primary (according to ``pg_stat_replication``)
+                       and are eligible to be listed in ``synchronous_standby_names``.
+        :param sync_wanted: desired number of synchronous nodes
+                            (``synchronous_node_count`` from the global configuration).
+        :param leader_wanted: the desired leader (could be different from the *leader* right after a failover).
+
+        """
+        self.leader = leader
+        self.quorum = quorum
+        self.voters = CaseInsensitiveSet(voters)
+        self.numsync = min(numsync, len(sync))  # numsync can't be bigger than number of listed synchronous nodes.
+        self.sync = CaseInsensitiveSet(sync)
+        self.numsync_confirmed = numsync_confirmed
+        self.active = CaseInsensitiveSet(active)
+        self.sync_wanted = sync_wanted
+        self.leader_wanted = leader_wanted
+
+    def check_invariants(self) -> None:
+        """Checks invatiant of ``synchronous_standby_names`` and ``/sync`` key in DCS.
+
+        :raises:
+            :exc:`QuorumError`: in case of broken state"""
+        voters = CaseInsensitiveSet(self.voters | CaseInsensitiveSet([self.leader]))
+        sync = CaseInsensitiveSet(self.sync | CaseInsensitiveSet([self.leader_wanted]))
+
+        # We need to verify that subset of nodes that can acknowledge a commit overlaps
+        # with any subset of nodes that can achieve quorum to promote a new leader.
+        if self.voters and not (len(voters | sync) <= self.quorum + self.numsync + 1):
+            raise QuorumError("Quorum and sync not guaranteed to overlap: nodes %d >= quorum %d + sync %d" %
+                              (len(voters | sync), self.quorum, self.numsync))
```

Comment:
```suggestion
            raise QuorumError(f"Quorum and sync not guaranteed to overlap: nodes {len(voters | sync)} > quorum "
                              f"{self.quorum} + sync {self.numsync} + 1.")
```

Use f-string and adjust the message to reflect the calculation.

Reply:
Perhaps, it's better to pre-define a var for readability.
```suggestion
            nodes = len(voters | sync)
            raise QuorumError(f"Quorum and sync not guaranteed to overlap: nodes {nodes} > quorum "
                              f"{self.quorum} + sync {self.numsync} + 1."))
```